doc: doxygen: fix remaining warnings with Doxygen v1.9.1

henrikbrixandersen opened the pull request:
Fix remaining Doxygen warnings when using Doxygen v1.9.1. Please see commit logs for further details.

henrikbrixandersen committed:
bluetooth: audio: cap: remove stray list item bullets

Remove stray list item bullets as they confuse Doxygen. Also fix wrong
Doxygen comment block intro.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: interrupt_controller: esp32: use backticks for code identifiers

Use double backticks for code identifiers as the wildcards confuse Doxygen.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: rtc: ds3231: use backticks for code identifiers

Use double backticks for code identifiers as the wildcards confuse Doxygen.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: i3c: fix nested list item formatting

The formatting of the nested list in the I3C header somehow confuses
Doxygen. Explicitly terminate each of the nested lists to avoid this.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
@gmarull Was the intention of https://github.com/zephyrproject-rtos/zephyr/commit/85a0e1325cc690155b95274725520434ca1484ff to catch Doxygen warnings in CI?

gmarull commented:
> @gmarull Was the intention of [85a0e13](https://github.com/zephyrproject-rtos/zephyr/commit/85a0e1325cc690155b95274725520434ca1484ff) to catch Doxygen warnings in CI?

Yes, and it should in fact work. The option you're changing is in fact patched by the Sphinx extension. Maybe there's an issue there? Does Sphinx really change something when tweaking that doxyfile option?

henrikbrixandersen commented:
> Yes, and it should in fact work.

It clearly doesn't, otherwise we'd be seeing documentation build failures everywhere. I removed a bunch of Doxygen warnings in #64964 and this PR removes the remaining ones. None of these create documentation build failures in CI.

> The option you're changing is in fact patched by the Sphinx extension. Maybe there's an issue there? Does Sphinx really change something when tweaking that doxyfile option?

I wouldn't know where to start looking, sorry.

gmarull commented:
> > Yes, and it should in fact work.
> 
> It clearly doesn't, otherwise we'd be seeing documentation build failures everywhere. I removed a bunch of Doxygen warnings in #64964 and this PR removes the remaining ones. None of these create documentation build failures in CI.
> 
> > The option you're changing is in fact patched by the Sphinx extension. Maybe there's an issue there? Does Sphinx really change something when tweaking that doxyfile option?
> 
> I wouldn't know where to start looking, sorry.

https://github.com/zephyrproject-rtos/zephyr/blob/91fdef0783e0355c85344617bbf202d7fe67831a/doc/_extensions/zephyr/doxyrunner.py#L250-L261

henrikbrixandersen commented:
So the build does fail locally, but it doesn't seem to be picked up in CI:

```
...
Running Sphinx v6.1.0
making output directory... done
Building Kconfig database...... done
Preparing Doxyfile...
Checking if Doxygen needs to be run...
Running Doxygen...
WARNING: [generated]:4: Invalid list item found
WARNING: [generated]:8: Invalid list item found
WARNING: [generated]:11: Invalid list item found
WARNING: [generated]:14: Invalid list item found
WARNING: [generated]:19: Invalid list item found
WARNING: [generated]:22: Invalid list item found
Syncing Doxygen output...
building [mo]: targets for 0 po files that are out of date
writing output... 
building [html]: targets for 1 source files that are out of date
updating environment: [new config] 1601 added, 0 changed, 0 removed
reading sources... [100%] snippets/xen_dom0/README                                                                                                                                                         
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] snippets/xen_dom0/README                                                                                                                                                          
generating indices... genindex done
writing additional pages... search done
copying images... [100%] ../../services/zbus/images/zbus_publishing_process_example2.svg                                                                                                                   
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
WARNING:root:Unused expression: .*Duplicate C declaration.*\n.*'\.\. c:.*:: .*struct in_addr.*'.*
WARNING:root:Unused expression: .*Duplicate C declaration.*\n.*'\.\. c:.*:: .*struct in6_addr.*'.*
build finished with problems, 6 warnings.
FAILED: CMakeFiles/html /home/hebad/Projects/zephyrproject/zephyr/doc/_build/CMakeFiles/html 
cd /home/hebad/Projects/zephyrproject/zephyr/doc/_build && /home/hebad/Projects/zephyrproject/venv/lib/python3.10/site-packages/cmake/data/bin/cmake -E env DOXYGEN_EXECUTABLE=/usr/bin/doxygen DOT_EXECUTABLE=/usr/bin/dot /home/hebad/Projects/zephyrproject/venv/bin/sphinx-build -b html -c /home/hebad/Projects/zephyrproject/zephyr/doc -d /home/hebad/Projects/zephyrproject/zephyr/doc/_build/doctrees -w /home/hebad/Projects/zephyrproject/zephyr/doc/_build/html.log -t development -j auto -W --keep-going -T -j1 /home/hebad/Projects/zephyrproject/zephyr/doc/_build/src /home/hebad/Projects/zephyrproject/zephyr/doc/_build/html
ninja: build stopped: subcommand failed.
make[1]: *** [Makefile:21: html] Error 1
make[1]: Leaving directory '/home/hebad/Projects/zephyrproject/zephyr/doc'
make: *** [Makefile:18: html-fast] Error 2

$ echo $? 
2

```

gmarull commented:
interesting... doxygen output must be redirected somewhere else :/ ? It used to work

kartben commented on include/zephyr/drivers/i3c.h:
```diff
@@ -35,27 +35,34 @@ extern "C" {
  *   - 1: I3C Controller capable
  *   - 2: Reserved
  *   - 3: Reserved
+ *   .
```
This is weird that this would be needed, which version of doxygen are you using? I can't seem to reproduce the warning? (1.9.6 here)

kartben replied:
CI is using 1.9.6. I don't think we necessarily "depend" on a newer version if the only difference is warnings in one and none in the other, but agreed this need to be clearer. Also there is also the opposite issue which is that IIRC generation fails for the most recent 1.9.8 (can't remember if this is "just" Breathe or in Doxygen proper)

gmarull replied:
Some Doxygen versions (e.g. 1.9.5) had bugs that made them unusable for Zephyr. Maybe 1.9.1 has some bugs as well.

henrikbrixandersen commented:
I have removed the change to the Doxygen configuration file and adjusted the description. Given that Ubuntu 22.04.3 LTS still ships with Doxygen v1.9.1, I think it would be nice to be able to eliminate the build warnings using that version still.

kartben commented on include/zephyr/drivers/rtc/maxim_ds3231.h:
```diff
@@ -15,7 +15,7 @@
  * The core Zephyr API to this device is as a counter, with the
  * following limitations:
- * * counter_read() and counter_*_alarm() cannot be invoked from
+ * * ``counter_read()`` and ``counter_*_alarm()`` cannot be invoked from
```
These are both APIs that don't exist anymore so it'd be nice to update this part, I guess. Maybe @pabigot can suggest an update as the original author?

kartben reviewed (approved):
Thanks!
Non-blocking comment as this is beyond the scope of this PR, but would definitely be nice to address.
